apps/global_full/4C_global_full_io.cpp

```diff
@@ -9,15 +9,185 @@
+      (arguments.restart_per_group.size() > 1 || arguments.restart_identifier_per_group.size() > 1))
+  {
+    FOUR_C_THROW(
+        "When using --nptype other than 'separateInputFiles', only one restart step and one "
+        "restartfrom identifier must be given.");
+  }
+
+  if (arguments.nptype != NPT::no_nested_parallelism && arguments.n_groups == 1)
+  {
+    FOUR_C_THROW("Number of groups must be greater than 1 when using nested parallelism.");
+  }
+
+  for (size_t i = 0; i < arguments.restart_identifier_per_group.size(); ++i)
+  {
+    if (i >= arguments.restart_per_group.size())
+    {
+      FOUR_C_THROW("You need to specify a restart step when using restartfrom.");
+    }
+  }
+}
+
+void adapt_legacy_cli_arguments(int* argc, char*** argv)
+{
+  std::vector<std::string> new_args{(*argv)[0]};  // program name
+  std::vector<std::string> pending_restarts, pending_restartfroms;
+
+  auto warn = [](const std::string& name, const std::string& to)
+  {
+    std::cerr << "DEPRECATION WARNING: Legacy  argument '" << name << "' has been converted to '"
+              << to
+              << "'. Please update your command line arguments. This legacy form will be removed "
+                 "in a future release.\n";
+  };
+
+  auto combine_and_warn =
+      [&warn](std::vector<std::string>& out_args, const char* name, std::vector<std::string>& vals)
+  {
+    if (!vals.empty())
+    {
+      std::string combined = std::string("--") + name + "=";
+      for (size_t i = 0; i < vals.size(); ++i)
+      {
+        if (i) combined += ",";
+        combined += vals[i];
+      }
+      out_args.push_back(combined);
+      warn(name, combined);
+      vals.clear();
+    }
+  };
+
+  for (int i = 1; i < *argc; ++i)
+  {
+    std::string arg = (*argv)[i];
+
+    // Collect dashless legacy restart arguments instead of pushing them immediately.
+    if (arg.rfind("restart=", 0) == 0)
+    {
+      pending_restarts.push_back(arg.substr(8));
+      continue;
+    }
+    if (arg.rfind("restartfrom=", 0) == 0)
+    {
+      pending_restartfroms.push_back(arg.substr(12));
+      continue;
+    }
+
+    // Convert known single-dash legacy options to double-dash
+    if (arg.rfind("-ngroup=", 0) == 0 || arg.rfind("-glayout=", 0) == 0 ||
+        arg.rfind("-nptype=", 0) == 0)
+    {
+      std::string new_arg = "-" + arg;
+      warn(arg, new_arg);
+      new_args.push_back(new_arg);
+      continue;
+    }
+
+    // Already a long option with two dashes: keep as-is
+    if (arg.rfind("--", 0) == 0)
+    {
+      new_args.push_back(arg);
+      continue;
+    }
+
+    // Keep everything else unchanged (positional args, -p, -h, etc.)
+    new_args.push_back(arg);
+  }
+
+  combine_and_warn(new_args, "restart", pending_restarts);
+  combine_and_warn(new_args, "restartfrom", pending_restartfroms);
+
+  *argc = static_cast<int>(new_args.size());
+  *argv = new char*[*argc];
+  for (int i = 0; i < *argc; ++i)
+  {
+    (*argv)[i] = strdup(new_args[i].c_str());
+  }
```

Comment:
I am almost certain you never `delete` the array. Let's do this differently: instead of modifying in-place, why not just return a `vector<string>` of the sanitized arguments? This is what you already have in this functions anyway.